include/permanent.hpp

```diff
@@ -322,34 +322,253 @@ std::complex<double> permanent_quad(std::vector<std::complex<double>> &mat) {
  */
-double permanent_quad(std::vector<double> &mat) {
+double permanent_quad(std::vector<double> &mat) 
+{
     std::vector<qp> matq(mat.begin(), mat.end());
     qp perm = permanent(matq);
     return static_cast<double>(perm);
 }
 
-
-
 /**
  * \rst
  *
- * Returns the permanent of a matrix using Ryser's algo with Gray code ordering with fsum
+ * Returns the permanent of a matrix using Ryser's algo 
+ * with Gray code ordering with fsum
  *
  * \endrst
  *
  *
- * This is a wrapper around the templated function `libwalrus::perm_fsum` for Python
- * integration. It accepts and returns double numeric types, and
- * returns sensible values for empty and non-even matrices.
+ * This is a wrapper around the templated function `libwalrus::perm_fsum` 
+ * for Python integration. It accepts and returns double numeric types, 
+ * and returns sensible values for empty and non-even matrices.
  *
  *
  * @param mat vector representing the flattened matrix
  * @return the permanent
  */
-double permanent_fsum(std::vector<double> &mat) {
+double permanent_fsum(std::vector<double> &mat) 
+{
     std::vector<double> matq(mat.begin(), mat.end());
     double perm = perm_fsum(matq);
     return static_cast<double>(perm);
 }
 
+
+/**
+ * Returns the permanent of a matrix (nthreads=1)
+ *
+ * \rst
+ *
+ * Returns the permanent of a matrix using the BBFG algorithm.
+ * This algorithm was given by Glynn (2010) with the time-complexity 
+ * of \f$O(n 2^n)\f$ using Gray code ordering.
+ *
+ * \endrst
+ *
+ *
+ * @param mat  a flattened vector of size \f$n^2\f$, representing an
+ *      \f$n\times n\f$ row-ordered symmetric matrix.
+ * @return permanent of the input matrix
+ */
+template <typename T>
+inline T perm_BBFG_serial(std::vector<T> &mat) 
+{
+    int n = std::sqrt(static_cast<double>(mat.size()));
+    llint x = static_cast<llint>(pow(2,n - 1));
+    std::vector<T> col_sum(n, static_cast<T>(0));
+
+    // init col_sum 
+    for (int i=0; i < n; ++i) {
+        for (int j=0; j < n; ++j) {
+            col_sum[i] += mat[j*n + i];
+        }
+    }
+
+    constexpr T p1 = static_cast<T>(1.0);
+    constexpr T p2 = static_cast<T>(2.0);
+    constexpr T n2 = static_cast<T>(-2.0);
+
+    T mulcolsum, coeff;
+    T total = static_cast<T>(0);
+    llint og=0, ng, gd;
+    int sgn=1, gdi;
+
+    // over all 2^{n-1} permutations of delta
+    for (llint k=1; k < x+1; ++k) {
+        mulcolsum = std::accumulate(
+                        col_sum.begin(), 
+                        col_sum.end(), 
+                        p1, 
+                        std::multiplies<T>());
+        total += sgn > 0 ? mulcolsum : -mulcolsum;
+        
+        // updating gray order
+        ng = igray(k);
+        gd = og ^ ng;
+
+        coeff = og > ng ? p2 : n2;
+        gdi = left_most_set_bit(gd);
+        for (int j=0; j < n; ++j) {
+            col_sum[j] += coeff * mat[gdi * n + j];
+        }
+
+        sgn = -sgn;
+        og = ng;
+    }
+
+    // TODO: overflow handling
```

Comment:
Any thoughts on the best approach here?

Reply:
The source of overflowing is the `std::pow` and casting to `T`. A common trick would be to use log to check the overflow from
```
2^{n-1} <= UPPER_BOUND iff n-1 <= log(UPPER_BOUND) / log(2)
```
However, the `std::pow` method itself takes care of such overflowing with returning `HUGE_VAL` (https://en.cppreference.com/w/cpp/numeric/math/pow). What do you think the best approach is here?

Reply:
I think keeping to the CPP standard library here and avoiding GNU libraries might be best for portability. I think a check for `HUGE_VAL` would be good here before returning, and abort if encountered.